lib/Dialect/Synth/Transforms/CutRewriter.cpp

```diff
@@ -0,0 +1,564 @@
+  Cut newCut;
+  // Topological sort the operations in the new cut.
+  // TODO: Merge-sort `operations` and `other.operations` by operation index
+  // (since it's already topo-sorted, we can use a simple merge).
+  std::function<void(Operation *)> populateOperations = [&](Operation *op) {
+    // If the operation is already in the cut, skip it
+    if (newCut.operations.contains(op))
+      return;
+
+    // Add its operands to the worklist
+    for (auto value : op->getOperands()) {
+      if (isAlwaysCutInput(value))
+        continue;
+
+      // If the value is in *both* cuts inputs, it is an input. So skip
+      // it.
+      bool isInput = inputs.contains(value);
+      bool isOtherInput = other.inputs.contains(value);
+      // If the value is in this cut inputs, it is an input. So skip it
+      if (isInput && isOtherInput)
+        continue;
+
+      auto *defOp = value.getDefiningOp();
+
+      assert(defOp && "Value must have a defining operation since block"
+                      "arguments are treated as inputs");
+
+      // Otherwise, check if the operation is in the other cut.
+
+      if (isInput) {
+        if (!other.operations.contains(defOp)) // op is in the other cut.
+          continue;
+      }
+      if (isOtherInput) {
+        if (!operations.contains(defOp)) // op is in this cut.
+          continue;
+      }
+
+      populateOperations(defOp);
+    }
+
+    // Add the operation to the cut
+    newCut.operations.insert(op);
+  };
+
+  populateOperations(root);
+
+  // A map for constructing the depth of each operation.
+  DenseMap<Operation *, unsigned> depthMap;
+
+  // Construct inputs.
+  for (auto *operation : newCut.operations) {
+    depthMap[operation] = 1;
+    for (auto value : operation->getOperands()) {
+      if (isAlwaysCutInput(value)) {
+        newCut.inputs.insert(value);
+        continue;
+      }
+
+      auto *defOp = value.getDefiningOp();
+      assert(defOp && "Value must have a defining operation");
+
+      depthMap[operation] = std::max(depthMap[operation], depthMap[defOp] + 1);
+      // If the operation is not in the cut, it is an input
+      if (!newCut.operations.contains(defOp))
+        // Add the input to the cut
+        newCut.inputs.insert(value);
+    }
+    unsigned depth = depthMap[operation];
+    newCut.depth = std::max(depth, newCut.depth);
+  }
+
+  // TODO: Sort the inputs by their defining operation.
+  // TODO: Update area and delay based on the merged cuts.
+
+  return newCut;
+}
+
+// Reroot the cut with a new root operation.
+// This is used to create a new cut with the same inputs and operations, but a
+// different root operation.
+Cut Cut::reRoot(Operation *root) const {
+  Cut newCut;
+  newCut.inputs = inputs;
+  newCut.operations = operations;
+  // Add the new root operation to the cut
+  newCut.operations.insert(root);
+  newCut.depth = depth + 1; // Increment depth for new root
+  return newCut;
+}
```

Comment:
Does this require that the old root op is an operand of the new root op? The `depth + 1` suggests that the added op must strictly extend the depth, which suggests that it must depend on the old root somehow. Should this possibly be some kind of assert?
```c++
assert(llvm::is_contained(root.getOperands(), operations.back()->getResult(0)));
```
Not sure how relevant this is.